providers/linode/linodeProvider.go

```diff
@@ -0,0 +1,303 @@
+		if err != nil {
+			return nil, err
+		}
+		j, err := json.Marshal(req)
+		if err != nil {
+			return nil, err
+		}
+		corr := &models.Correction{
+			Msg: fmt.Sprintf("%s: %s", m.String(), string(j)),
+			F: func() error {
+				record, err := api.createRecord(domainID, req)
+				if err != nil {
+					return err
+				}
+				// TTL isn't saved when creating a record, so we will need to modify it immediately afterwards
+				return api.modifyRecord(domainID, record.ID, req)
+			},
+		}
+		corrections = append(corrections, corr)
+	}
+	for _, m := range modify {
+		id := m.Existing.Original.(*domainRecord).ID
+		if id == 0 { // Skip ID 0, these are the default nameservers always present
+			continue
+		}
+		req, err := toReq(dc, m.Desired)
+		if err != nil {
+			return nil, err
+		}
+		j, err := json.Marshal(req)
+		if err != nil {
+			return nil, err
+		}
+		corr := &models.Correction{
+			Msg: fmt.Sprintf("%s, Linode ID: %d: %s", m.String(), id, string(j)),
+			F: func() error {
+				return api.modifyRecord(domainID, id, req)
+			},
+		}
+		corrections = append(corrections, corr)
+	}
+
+	return corrections, nil
+}
+
+func toRc(dc *models.DomainConfig, r *domainRecord) *models.RecordConfig {
+	// This handles "@" etc.
+	name := dnsutil.AddOrigin(r.Name, dc.Name)
+
+	target := r.Target
+	// Make target FQDN (#rtype_variations)
+	if r.Type == "CNAME" || r.Type == "MX" || r.Type == "NS" || r.Type == "SRV" {
+		target = dnsutil.AddOrigin(target+".", dc.Name)
+	}
+
+	return &models.RecordConfig{
+		NameFQDN:     name,
+		Type:         r.Type,
+		Target:       target,
+		TTL:          uint32(r.TTLSec),
+		MxPreference: uint16(r.Priority),
+		SrvPriority:  uint16(r.Priority),
+		SrvWeight:    uint16(r.Weight),
+		SrvPort:      uint16(r.Port),
+		Original:     r,
+	}
+}
+
+func toReq(dc *models.DomainConfig, rc *models.RecordConfig) (*recordEditRequest, error) {
+	req := &recordEditRequest{
+		Type:     rc.Type,
+		Name:     dnsutil.TrimDomainName(rc.NameFQDN, dc.Name),
+		Target:   rc.Target,
+		TTL:      int(rc.TTL),
+		Priority: 0,
+		Port:     int(rc.SrvPort),
+		Weight:   int(rc.SrvWeight),
+	}
+
+	// Linode doesn't use "@", it uses an empty name
+	if req.Name == "@" {
+		req.Name = ""
+	}
+
+	// Linode uses the same property for MX and SRV priority
+	switch rc.Type { // #rtype_variations
+	case "MX":
+		req.Priority = int(rc.MxPreference)
+		req.Target = fixTarget(req.Target, dc.Name)
+	case "SRV":
+		req.Priority = int(rc.SrvPriority)
+
+		// From softlayer provider
+		// This is to support SRV, it doesn't work yet for Linode
+		result := srvRegexp.FindStringSubmatch(req.Name)
+
+		if len(result) != 3 {
+			return nil, fmt.Errorf("SRV Record must match format \"_service._protocol\" not %s", req.Name)
+		}
+
+		var serviceName, protocol string = result[1], strings.ToLower(result[2])
+
+		req.Protocol = protocol
+		req.Service = serviceName
+		req.Name = ""
+	case "CNAME":
+		req.Target = fixTarget(req.Target, dc.Name)
+	}
```

Comment:
Switch statements should list all possible rtypes so that when a new rtype is added, we fail fast.  Look at https://github.com/StackExchange/dnscontrol/blob/master/models/dns.go#L90 for an example.